services/comments.js

```diff
@@ -51,50 +51,76 @@ module.exports = class CommentsService {
 
   /**
    * Edit a Comment
-   * @param {CommentModel|String} comment you want to edit (or its ID)
-   * @param {String} body  the new Comment body
-   * @param {String} status  the new Comment status
+   * @param {String} id    comment.id you want to edit (or its ID)
+   * @param {String} asset_id   asset_id of the comment
+   * @param {String} editor     user.id of the user trying to edit the comment (will err if not comment author)
+   * @param {String} body       the new Comment body
+   * @param {String} status     the new Comment status
    */
-  static async edit(comment, {body, status, ignoreEditWindow}) {
-    if (typeof comment === 'string') {
-
-      // it's an id
-      comment = await this.findById(comment);
-    }
-    const editWindowExpired = (comment) => {
-      const now = new Date;
-      const editableUntil = this.getEditableUntilDate(comment);
-      return now > editableUntil;
-    };
-    if (( ! ignoreEditWindow) && editWindowExpired(comment)) {
-      throw Object.assign(new Error('Edit window is over.'), {
-        name: 'EditWindowExpired'
-      });
-    }
+  static async edit(id, asset_id, editor, {body, status, ignoreEditWindow}) {
     if (status && ! STATUSES.includes(status)) {
       throw new Error(`status ${status} is not supported`);
     }
-
-    const {id} = comment;
-    const {nModified} = await CommentModel.update({id}, {
-      $set: {
-        body,
-        status,
+    const lastEditableCommentCreatedAt = new Date((new Date()).getTime() - EDIT_WINDOW_MS);
+    const filter = Object.assign(
+      {
+        id,
+        asset_id,
+        author_id: editor,
       },
-      $push: {
-        body_history: {
+      ignoreEditWindow ? {} : {
+        created_at: {
+          $gt: lastEditableCommentCreatedAt,
+        },
+      }
+    );
+    const {nModified} = await CommentModel.update(
+      filter,
+      {
+        $set: {
           body,
-          created_at: new Date(),
+          status,
         },
-        status_history: {
-          type: status,
-          created_at: new Date(),
-        }
-      },
-    });
+        $push: {
+          body_history: {
+            body,
+            created_at: new Date(),
+          },
+          status_history: {
+            type: status,
+            created_at: new Date(),
+          }
+        },
+      }
+    );
     switch (nModified) {
-    case 0:
-      throw new Error(`Couldn't edit comment. There is no Comment with id "${id}"`);
+    case 0: {
+
+          // disambiguate possible error cases
+      const comment = await this.findById(id);
```

Comment:
From my perspective, any path that does not aggressively minimize the number of queries being executed against the database is one that needs reconsideration.

The 'cost' may be a few ms per request, but consider a high traffic site where that number can easily explode to a significant load against the database server.

Ideally, when faced with any one function, we should keep these to an absolute minimum, justifying each and every external service call. We should definitely return useful error messages, but not at the expense of slowing down the system, that's why I've always advocated use of defensive database queries that negate the need for non-atomic style operations which would fail spectacularly with a non-transaction based database like Mongo.

Instead of using an `update` operation, mongoose (through the underlying MongoDB driver) provides a `findOneAndUpdate` which would reduce the query cost down to 1 rather than 2.